### ISSUE TITLE
quotients for ZZ do not get routed to RingElement

### ISSUE THREAD

**Issue by mahrud** [Macaulay2/M2#1519]
Observation:
```m2
i1 : ideal 2 : ideal 6

o1 = ideal 1

o1 : Ideal of ZZ

i2 : ideal 4 : ideal 6
stdio:2:9:(3): error: no method found for applying quotient to:
     argument 1 :  ideal 4 (of class Ideal)
     argument 2 :  6 (of class ZZ)
```
The reason the first one works is a quick containment check in quotient, but when there's no containment, `quotient(ideal 2, 6)` is called, which should be picked up by `(quetient, Ideal, RingElement)`, but for some reason RingElement is not in the ancestors of ZZ:
```m2
i3 : ancestors ZZ

o3 = {ZZ, Number, Thing}
```
Is this on purpose?

**Comment by mahrud** [comment]
I suspect this is also the reason why something like this doesn't work:
```m2
i1 : map(ZZ, ZZ, 2)
stdio:1:1:(3): error: no method found for applying map to:
     argument 1 :  ZZ (of class Ring)
     argument 2 :  ZZ (of class Ring)
     argument 3 :  2 (of class ZZ)
```

**Comment by DanGrayson** [comment]
>  for some reason RingElement is not in the ancestors of ZZ:
> 
> ```m2
> i3 : ancestors ZZ
> 
> o3 = {ZZ, Number, Thing}
> ```
> 
> Is this on purpose?

Yes, the objects of type RingElement are handled by the engine and are polynomials.

**Comment by DanGrayson** [comment]
> I suspect this is also the reason why something like this doesn't work:
> 
> ```m2
> i1 : map(ZZ, ZZ, 2)
> stdio:1:1:(3): error: no method found for applying map to:
>      argument 1 :  ZZ (of class Ring)
>      argument 2 :  ZZ (of class Ring)
>      argument 3 :  2 (of class ZZ)
> ```

There's only one ring homomorphism from ZZ to ZZ, so `map(ZZ, ZZ, 2)` makes no sense and shouldn't work.

**Comment by mahrud** [comment]
> Yes, the objects of type RingElement are handled by the engine and are polynomials.

This doesn't make any sense. ZZ is a ring, and its elements are handled by flint and gmp.
```m2
i1 : R = ZZ[]

o1 = R

o1 : PolynomialRing

i2 : ideal 4_R : ideal 6_R

o2 = ideal 2

o2 : Ideal of R
```

> There's only one ring homomorphism from ZZ to ZZ, so `map(ZZ, ZZ, 2)` makes no sense and shouldn't work.

That's completely beside the point:
```m2
i4 : map(ZZ, ZZ, 1)
stdio:4:1:(3): error: no method found for applying map to:
     argument 1 :  ZZ (of class Ring)
     argument 2 :  ZZ (of class Ring)
     argument 3 :  1 (of class ZZ)

i6 : map(ZZ, ZZ, 0)
stdio:6:1:(3): error: no method found for applying map to:
     argument 1 :  ZZ (of class Ring)
     argument 2 :  ZZ (of class Ring)
     argument 3 :  0 (of class ZZ)
```

This sort of stuff should work automatically, so numbers should become RingElements.

**Comment by DanGrayson** [comment]
> > Yes, the objects of type RingElement are handled by the engine and are polynomials.
> 
> This doesn't make any sense. ZZ is a ring, and its elements are handled by flint and gmp.

Not all desired mathematical concepts are supported, because those things also have different implementations.  Ring elements are basic lists, and integers are not.

> 
> ```m2
> i1 : R = ZZ[]
> 
> o1 = R
> 
> o1 : PolynomialRing
> 
> i2 : ideal 4_R : ideal 6_R
> 
> o2 = ideal 2
> 
> o2 : Ideal of R
> ```
> 
> > There's only one ring homomorphism from ZZ to ZZ, so `map(ZZ, ZZ, 2)` makes no sense and shouldn't work.
> 
> Two, actually, and if that's your reason then it should give an error. In any case though, that's completely beside the point:
> 
> ```m2
> i4 : map(ZZ, ZZ, 1)
> stdio:4:1:(3): error: no method found for applying map to:
>      argument 1 :  ZZ (of class Ring)
>      argument 2 :  ZZ (of class Ring)
>      argument 3 :  1 (of class ZZ)
> 
> i6 : map(ZZ, ZZ, 0)
> stdio:6:1:(3): error: no method found for applying map to:
>      argument 1 :  ZZ (of class Ring)
>      argument 2 :  ZZ (of class Ring)
>      argument 3 :  0 (of class ZZ)
> ```
> 
> This sort of stuff should work automatically, so numbers should become RingElements.

There's only one, because ring homomorphisms send 1 to 1.

**Comment by mahrud** [comment]
> Not all desired mathematical concepts are supported, because those things also have different implementations. Ring elements are basic lists, and integers are not.

Then RingElement should be renamed to something else.

> There's only one, because ring homomorphisms send 1 to 1.

You can send everything to zero.

**Comment by DanGrayson** [comment]
> > Not all desired mathematical concepts are supported, because those things also have different implementations. Ring elements are basic lists, and integers are not.
> 
> Then RingElement should be renamed to something else.

Let's stick with the current name.

> 
> > There's only one, because ring homomorphisms send 1 to 1.
> 
> You can send everything to zero.

No, you can't, because ring homomorphisms send 1 to 1.

**Comment by mahrud** [comment]
This sort of resistance just makes life harder for programmers, because you have to add extra checks and conditions to distinguish "RingElements" from "ring elements". Is there no interest in making basic constructions work categorically?

A simple case of how annoying this is:
```m2
i1 : ideal 2 : ideal 6

o1 = ideal 1

o1 : Ideal of ZZ

i2 : ideal 4 : ideal 6
stdio:2:9:(3): error: no method found for applying quotient to:
     argument 1 :  ideal 4 (of class Ideal)
     argument 2 :  6 (of class ZZ)
```

**Comment by DanGrayson** [comment]
Sure, improvements are always welcome.  What's your proposal?

**Comment by mikestillman** [comment]
It is true that this is a bit harder for programmers.  However, we want to consider integers as just arbitrary precision integers, not going through the back end at all.  We have tried to make it transparent for the user.  Obviously we have missed some things, and it would be great to get all of these operations working over ZZ and polynomial rings over ZZ.  But making everything (e.g. an integer) an element in an engine ring would cause a performance hit that seems bad.

However, if you have a suggestion for improving the situation , I'd like to hear it!

**Comment by mikestillman** [comment]
@mahrud This particular code you mention is certainly a bug.

**Comment by mahrud** [comment]
I mean, I'm not proposing a change in the engine here, just a change in the toplevel inheritence structure so at least ZZ and RingElement have a reasonable common ancestor.

Is there an example of a toplevel function that should work for RingElement but explicitly not ZZ elements? We could change the signature of such functions to be EngineRingElement, whose common ancestor with ZZ is RingElement perhaps.

**Comment by DanGrayson** [comment]
If inheritance were used just for mathematical aspects of objects, then something like that might work, but inheritance is also used for datatype aspects of objects.   In this case, a ring element is actually a basic list, whereas an integer is actually a pointer to a struct of type ZZcell, which contains a pointer to a gmp struct of type mpz_srcptr.

**Comment by mahrud** [comment]
> If inheritance were used just for mathematical aspects of objects, then something like that might work, but inheritance is also used for datatype aspects of objects. In this case, a ring element is actually a basic list, whereas an integer is actually a pointer to a struct of type ZZcell, which contains a pointer to a gmp struct of type mpz_srcptr.

The solution would be to move the common ancestor far back. I looked into this, but couldn't find where RingElement is defined! Why is it a BasicList? In particular, are there any methods installed for BasicList that we want to make work with RingElement?

**Comment by DanGrayson** [comment]
It's a basic list so it can contain a pointer to the raw ring element.

**Comment by mahrud** [comment]
> It's a basic list so it can contain a pointer to the raw ring element.

Don't both of them contain a pointer in the same way?
```m2
i78 : raw 2

o78 = 2

o78 : RawRingElement

i79 : raw x

o79 = x

o79 : RawRingElement
```
Why would one require being a BasicList but not the other?

**Comment by DanGrayson** [comment]
No, `raw 2` does a conversion:
```m2
i10 : code (raw, ZZ)

o10 = -- code for method: raw(ZZ)
      ../m2/integers.m2:13:13-13:39: --source code:
      raw ZZ := x -> rawFromNumber(rawZZ(), x)
```

**Comment by mahrud** [comment]
You said both of them contain pointers, one to a gmp integer and one to a ring element. What's the difference?

Can RingElement contain more than one element?

**Comment by DanGrayson** [comment]
> You said both of them contain pointers, one to a gmp integer and one to a ring element. What's the difference?

Maybe what you want to propose is that ZZ should be re-ndefined as `new Type of RingElement` and that an object of type ZZ should contain a pointer to a ZZcell as its single entry.

> 
> Can RingElement contain more than one element?

An object of type RingElement always has just one entry.

**Comment by mahrud** [comment]
> Maybe what you want to propose is that ZZ should be re-ndefined as `new Type of RingElement` and that an object of type ZZ should contain a pointer to a ZZcell as its single entry.

This was my suggestion from the beginning. Both Number and EngineRingElement can inherit from RingElement, with the former containing a pointer to a ZZcell and latter containing a pointer to an engine ring element. Then toplevel functions can deal only with RingElement, unless they want to distinguish the two from each other.

> An object of type RingElement always has just one entry.

I'm confused about this: why have a list of just one entry instead of it being just a pointer type?

**Comment by DanGrayson** [comment]
> > An object of type RingElement always has just one entry.
> 
> I'm confused about this: why have a list of just one entry instead of it being just a pointer type?

Macaulay2's top level language doesn't have "pointer types".

**Comment by DanGrayson** [comment]
> > You said both of them contain pointers, one to a gmp integer and one to a ring element. What's the difference?
> 
> Maybe what you want to propose is that ZZ should be re-ndefined as `new Type of RingElement` and that an object of type ZZ should contain a pointer to a ZZcell as its single entry.
> 

There would be a performance hit, as there would be another memory allocation to make that list.

**Comment by mahrud** [comment]
> PS: There would be a performance hit, as there would be another memory allocation to make that list.

I keep asking, why are they a list at all? there seems to be no point in having a list at all.

> Macaulay2's top level language doesn't have "pointer types".

I never said top level. ZZcell is a Pointer type, why can't RingElement be one?

**Comment by DanGrayson** [comment]
> > PS: There would be a performance hit, as there would be another memory allocation to make that list.
> 
> I keep asking, why are they a list at all? there seems to be no point in having a list at all.

Another reason to make it a list is so you have a place to put the pointer to the class of the object (the ring of the ring element).  It also can save time to pre-computer the hash code of the ring element and save it.

> 
> > Macaulay2's top level language doesn't have "pointer types".
> 
> I never said top level. ZZcell simply contains a ZZ or ZZmutable, both of which are of Pointer types, why can't RingElement be one?

No, ZZcell contains a ZZ.  It's not enough for a top level ring element to have just a pointer to the raw ring element -- you also need a pointer to the top level ring that it's in.

**Comment by mahrud** [comment]
Ah, sorry I thought ZZcell could also be mutable. Probably it should though, shouldn't it?

> No, ZZcell contains a ZZ. It's not enough for a top level ring element to have just a pointer to the raw ring element -- you also need a pointer to the top level ring that it's in.

I'm still confused. Here are the two definitions:
```m2
export RawRingElement := Pointer "const struct RingElement *";
export ZZ := Pointer "mpz_srcptr";
```
From what I understand, `ring RingElement` ends up here:
https://github.com/Macaulay2/M2/blob/7ba26782adc33cb23f4ca3522130721f53b93fb8/M2/Macaulay2/d/interface.dd#L1370-L1378
So the ring is part of the pointer itself. In particular, both RingElement and ZZ end up here. How does the list come into play?

It might help if you show me an instance where the list has more than just a pointer, and possibly a way to get at those objects.

**Comment by DanGrayson** [comment]
> Ah, sorry I thought ZZcell could also be mutable. Probably it should though, shouldn't it?

No, once an integer reaches top level it should be libgc memory, to avoid memory leaks.

> 
> > No, ZZcell contains a ZZ. It's not enough for a top level ring element to have just a pointer to the raw ring element -- you also need a pointer to the top level ring that it's in.
> 
> I'm still confused. Here are the two definitions:
> 
> ```m2
> export RawRingElement := Pointer "const struct RingElement *";
> export ZZ := Pointer "mpz_srcptr";
> ```
> 
> From what I understand, `ring RingElement` ends up here:
> https://github.com/Macaulay2/M2/blob/7ba26782adc33cb23f4ca3522130721f53b93fb8/M2/Macaulay2/d/interface.dd#L1370-L1378
> 

No, that's the code for the the ring of a raw ring element, which is a raw ring, not a top level ring.

> 
> So the ring is part of the pointer itself. In particular, both RingElement and ZZ end up here. How does the list come into play?

 so you have a place to put the pointer to the class of the object (the ring of the ring element). It also can save time to pre-compute the hash code of the ring element and save it.



> It might help if you show me an instance where the list has more than just a pointer, and possibly a way to get at those objects.

The definition shows what it has:
```
export List := {+
     Class:HashTable,
     v:Sequence,
     hash:int,
     Mutable:bool
     };
```

**Comment by mahrud** [comment]
I see, so RawRingElement and RingElement are completely separate.

The confusion persists: `ringElementClass` is never used in `d/`, only defined and hooked up to `RingElement`. Can you show me where the fact that RingElement is a BasicList is used in the toplevel? I can't find any references, not even where hash is computed or references for it.

**Comment by mahrud** [comment]
To clarify, I know how `class` or `hash` are defined, but I don't see why those can't be cached by the engine and recalled by the interpreter.

**Comment by DanGrayson** [comment]
> I see, so RawRingElement and RingElement are completely separate.
> 
> The confusion persists: `ringElementClass` is never used in `d/`, only defined and hooked up to `RingElement`. Can you show me where the fact that RingElement is a BasicList is used in the toplevel? I can't find any references, not even where hash is computed or references for it.

For example, here:
```m2
new RingElement from RawRingElement := (R, f) -> (
     -- this might take too much time:
     -- if R.RawRing =!= rawRing f then error "internal error: raw ring mismatch encountered";
     new R from {f})
```
and here:
```m2
raw RingElement := f -> f#0
```

**Comment by DanGrayson** [comment]
> To clarify, I know how `class` or `hash` are defined, but I don't see why those can't be cached by the engine and recalled by the interpreter.

The engine doesn't know about top level rings, so it can't cache it.  The order of events at top level, when making a ring, is to make a raw ring and then put it inside a top level ring.

The engine could cache hash codes, but it doesn't.  Why bother, since they are used only at top level?

**Comment by mahrud** [comment]
> Why bother, since they are used only at top level?

The obvious reason is to make the type hierarchy more accurate, so ZZ and RingElement can have a reasonable common ancestor. I think making RingElement inherit from a simpler object than BasicList is probably the first step.

**Comment by DanGrayson** [comment]
> > Why bother, since they are used only at top level?
> 
> The obvious reason is to make the type hierarchy more accurate, so ZZ and RingElement can have a reasonable common ancestor. I think making RingElement inherit from a simpler object than BasicList is probably the first step.

Something like that might work, but if that's your aim, the hash code of a ring element could still be cached by the interpreter and not by the engine.  The engine doesn't want to waste its time on that, since not all ring elements find their way to top level.

**Comment by DanGrayson** [comment]
I added a few method definitions, so now the error will is different:
```m2
i1 : ideal 4 : ideal 6
stdio:1:9:(3): error: ZZ/n not implemented yet for composite n = 4
```

**Comment by DanGrayson** [comment]
The current algorithm that uses the quotient ring can probably be rewritten to not use it, as this code illustrates:
```m2
i1 : ideal ( module ideal 4 : ideal 6 )

o1 = ideal 2

o1 : Ideal of ZZ
```

**Comment by mahrud** [comment]
Actually, your commit is going to cause a conflict with our ongoing work in #1385. Could you revert it please? The issue I raised above is not specific to quotient, it's about the fact that numbers are not treated as ring elements.

**Comment by DanGrayson** [comment]
done

**Comment by mahrud** [comment]
Is there a method for retrieving a method function by looking in the ancestors of the input classes until a match is found? For instance, I'm having trouble figuring out how `new TEX from 2` works:
```m2
i2 : new TEX from 2

o2 = 2

o2 : TEX

i3 : class 2

o3 = ZZ

o3 : Ring

i4 : ancestors oo

o4 = {ZZ, Number, Thing}

o4 : List

i5 : methods (NewFromMethod, TEX)

o5 = {0 => (NewFromMethod, TEX, BasicList)}
     {1 => (NewFromMethod, TEX, String)   }

o5 : NumberedVerticalList
```

Neither of the listed methods should apply to "2", so how does this work?

**Comment by DanGrayson** [comment]
Yes, use "lookup".

```m2
i33 : code lookup(NewFromMethod, TEX, class 2)

o33 = ../src/M2/M2.git/M2/Macaulay2/m2/hypertext.m2:25:36-25:40: --source code:
      new Hypertext from Thing  := (M,x) -> {x}
```

**Comment by d-torrance** [comment]
> **Written by Claude** (Claude Opus 5, via Claude Code), not by @d-torrance, whose account posted it -- please weigh it accordingly.

The failure reported here is fixed, though not in the way the title asks for.

```m2
i1 : ideal 4 : ideal 6

o1 = ideal 2
```

which is right — `{n : 4 | 6n} = (2)`.

The fix is [`549828338a`](https://github.com/Macaulay2/M2/commit/549828338a) (2020-11-23,
*"doc updates in Colon"*), which added a `Number` overload beside the existing one:

```m2
quotient(Ideal,  Number)      :=
quotient(Ideal,  RingElement) := Ideal  => opts -> (I, f) -> quotient(I, ideal f, opts)
```

now at `Saturation.m2:146-147`, after the `Colon` package was renamed. It reached `development`
through [#1385](https://github.com/Macaulay2/M2/pull/1385) — worth noting that branch was opened
before this issue was filed and merged after it, so the PR's own dates are misleading; the commit
is the thing to look at.

### The question in the title is still open

The routing was not changed. `ancestors ZZ` is unchanged:

```m2
i2 : ancestors ZZ

o2 = {ZZ, Number, Thing}
```

So `RingElement` still is not an ancestor of `ZZ`, and `quotient(Ideal, RingElement)` still does not
catch integers — a parallel `Number` method was added instead. That answers the symptom without
answering *"Is this on purpose?"*, which is what you actually asked.

If that question is worth keeping alive, [#1477](https://github.com/Macaulay2/M2/issues/1477) is
the better home: it is your same question about the dispatch of `lift` and `promote`, and it is
still open.

<sub>Triaged in a sweep of the open issues; source of truth: [`bug-triage/issues.tsv`](https://github.com/d-torrance/M2/blob/bug-triage/bug-triage/issues.tsv).</sub>
